python/taichi/examples/simulation/implicit_fem.py

```diff
@@ -0,0 +1,353 @@
+    add(x, x, dt, v)
+
+
+@ti.kernel
+def advect():
+    for p in x:
+        v[p] += dt * (f[p] / m[p])
+        x[p] += dt * v[p]
+        f[p] = ti.Vector([0, 0, 0])
+
+
+@ti.kernel
+def init():
+    for u in x:
+        x[u] = ox[u]
+        v[u] = [0.0] * 3
+        f[u] = [0.0] * 3
+        m[u] = 0.0
+    for c in vertices:
+        F = Ds(vertices[c])
+        B[c] = F.inverse()
+        W[c] = ti.abs(F.determinant()) / 6
+        for i in range(4):
+            m[vertices[c][i]] += W[c] / 4 * density
+    for u in x:
+        x[u].y += 1.0
+
+
+@ti.kernel
+def floor_bound():
+    for u in x:
+        if x[u].y < 0:
+            x[u].y = 0
+            if v[u].y < 0:
+                v[u].y = 0
+
+
+@ti.func
+def check(u):
+    ans = 0
+    rest = u
+    for i in ti.static(range(3)):
+        k = rest % n_cube[2 - i]
+        rest = rest // n_cube[2 - i]
+        if k == 0: ans |= (1 << (i * 2))
+        if k == n_cube[2 - i] - 1: ans |= (1 << (i * 2 + 1))
+    return ans
+
+
+su = 0
+for i in range(3):
+    su += (n_cube[i] - 1) * (n_cube[(i + 1) % 3] - 1)
+indices = ti.field(ti.i32, shape=2 * su * 2 * 3)
+
+
+@ti.kernel
+def get_indices():
+    # calculate all the meshes on surface
+    cnt = 0
+    for c in vertices:
+        if c % 5 != 4:
+            for i in ti.static([0, 2, 3]):
+                verts = [vertices[c][(i + j) % 4] for j in range(3)]
+                sum = check(verts[0]) & check(verts[1]) & check(verts[2])
+                if sum:
+                    m = ti.atomic_add(cnt, 1)
+                    det = ti.Matrix.rows([
+                        x[verts[i]] - [0.5, 1.5, 0.5] for i in range(3)
+                    ]).determinant()
+                    if det < 0:
+                        tmp = verts[1]
+                        verts[1] = verts[2]
+                        verts[2] = tmp
+                    indices[m * 3] = verts[0]
+                    indices[m * 3 + 1] = verts[1]
+                    indices[m * 3 + 2] = verts[2]
+
+
+def substep():
+    if args.exp == 'explicit':
+        for i in range(10):
+            get_force()
+            advect()
+    else:
+        for i in range(1):
+            cg()
+    floor_bound()
+
+
+get_vertices()
+init()
+get_indices()
+
+if args.gui == 'ggui':
+    res = (800, 600)
+    window = ti.ui.Window("Implicit FEM", res, vsync=True)
+
+    frame_id = 0
+    canvas = window.get_canvas()
+    scene = ti.ui.Scene()
+    camera = ti.ui.make_camera()
+    camera.position(2.0, 2.0, 3.95)
+    camera.lookat(0.5, 0.5, 0.5)
+    camera.fov(55)
+
+    def render():
+        camera.track_user_inputs(window,
+                                 movement_speed=0.03,
+                                 hold_key=ti.ui.RMB)
+        scene.set_camera(camera)
+
+        scene.ambient_light((0.1, ) * 3)
+
+        scene.point_light(pos=(0.5, 10.0, 0.5), color=(0.5, 0.5, 0.5))
+        scene.point_light(pos=(10.0, 10.0, 10.0), color=(0.5, 0.5, 0.5))
+
+        #scene.particles(x, radius=0.02)
```

Comment:
nit: we'd better not leave commented code

Reply:
I will fix this.